jq/jq2.go

```diff
@@ -0,0 +1,453 @@
+		return jvArray(val)
+	case reflect.Map:
+		return jvMap(val)
+	}
+
+	panic("unknown type attempted to be cast to jv")
+}
+
+func jvMap(val reflect.Value) C.jv {
+	jvObj := C.jv_object()
+
+	for _, key := range val.MapKeys() {
+		// These allocations are freed when the whole object is freed.
+		keyJv := jvString(key.String())
+		valJv := jvInterface(val.MapIndex(key).Interface())
+		C.jv_object_set(jvObj, keyJv, valJv)
+	}
+
+	return jvObj
+}
+
+func jvArray(val reflect.Value) C.jv {
+	len := val.Len()
+	jvArray := C.jv_array_sized(C.int(len))
+
+	for i := 0; i < len; i++ {
+		// These allocations are freed when the whole array is freed.
+		jvArray = C.jv_array_set(jvArray, C.int(i), jvInterface(val.Index(i).Interface()))
+	}
+
+	return jvArray
+}
+
+func jvString(i interface{}) C.jv {
+	str := i.(string)
+	cstr := C.CString(str)
+	defer C.free(unsafe.Pointer(cstr))
+	return C.jv_string_sized(cstr, C.int(len(str)))
+}
+
+func jvNumber(i interface{}) C.jv {
+	val := reflect.ValueOf(i)
+	switch val.Kind() {
+	case reflect.Float32, reflect.Float64:
+		return C.jv_number(C.double(val.Float()))
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return C.jv_number(C.double(float64(val.Int())))
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return C.jv_number(C.double(float64(val.Uint())))
+	}
+
+	panic("unknown type for go number")
+}
+
+// interfaceValue maps an empty interface into a JSON Value.
+func interfaceValue(i interface{}) value {
+	jv := jvInterface(i)
+	defer C.jv_free(jv)
+	return jvToValue(jv)
+}
+
+var (
+	// libjq uses callbacks for error handling.
+	// This map safely stores errors under a key for a particular call.
+	// See https://github.com/golang/go/wiki/cgo#function-variables
+	callbackErrors  = make(map[uint64]chan error)
+	callbackErrorsM sync.RWMutex
+)
+
+// Session represents a handle to internal jq state.
+type Session struct {
+	state *C.struct_jq_state
+}
+
+// NewSession allocates a new jq state.
+func NewSession() (*Session, error) {
+	state, err := C.jq_init()
+	if err != nil {
+		return nil, err
+	} else if state == nil {
+		panic("failed to initialize jq state")
+	}
+
+	return &Session{state}, nil
+}
+
+// Close tears down any leftover jq state.
+func (s *Session) Close() {
+	// jq_teardown takes a **jq_state, thus the reference.
+	C.jq_teardown(&s.state)
+	s.state = nil
+}
+
+func errorFromJV(jv C.jv) error {
+	jv = C.jq_format_error(jv)
+	defer C.jv_free(jv)
+
+	if C.jv_get_kind(jv) == C.JV_KIND_NULL {
+		return nil
+	}
+	return errors.New(C.GoString(C.jv_string_value(jv)))
+}
+
+//export goJQErrorHandler
+func goJQErrorHandler(id uint64, jv C.jv) {
+	callbackErrorsM.Lock()
+	defer callbackErrorsM.Unlock()
+
+	err := errorFromJV(jv)
+	if err == nil {
+		return
+	}
+
+	callbackErrors[id] <- err
+}
+
+type StringerFunc func([]fmt.Stringer) error
+
+func (s *Session) Exec(program string, args, input interface{}, fn StringerFunc) error {
```

Comment:
I think this should become a regular function that constructs a session, calls ExecuteProgram, calls `fn`, then closes the session, and returns. As it is, it's possible to call close prior to `Exec` so I don't see this function providing any utility over the regular `session.ExecuteProgram` method.

Reply:
```suggestion
func Exec(program string, args, input interface{}, fn StringerFunc) error {
        s, err := NewSession()
	if err != nil {
		return err
	}
         defer s.Close() 
	values, err := s.ExecuteProgram(program, interfaceValue(args), interfaceValue(input))
	if err != nil {
		return err
	}

	stringers := make([]fmt.Stringer, len(values))
	for _, value := range values {
		stringers = append(stringers, value)
	}

	return fn(stringers)
}
```